haplotype_tracking.md

```diff
@@ -0,0 +1,442 @@
+    
+    def __repr__(self):
+        return self.segments.__repr__()
+    
+    def remove_segment(self, left, right):
+        """
+        Removes the intersection with [left, right).
+        """
+        removed = []
+        new = []
+        for a, b in self.segments:
+            u, v = a, min(b, left)
+            if u < v:
+                new.append([u, v])
+            u, v = max(a, left), min(b, right)
+            if u < v:
+                removed.append([u, v])
+            u, v = max(a, right), b
+            if u < v:
+                new.append([u, v])
+        self.segments = new
+        return SegmentList(removed)
+    
+    def add_segment(self, left, right):
+        """
+        Updates to the union with [left, right).
+        """
+        new = []
+        for a, b in self.segments:
+            if max(a, left) < min(b, right):
+                # overlaps
+                left = min(left, a)
+                right = max(right, b)
+            else:
+                new.append([a, b])
+        new.append([left, right])
+        self.segments = new
+```
+
+Here is a small tree sequence:
+
+```{code-cell} ipython3
+:tags: [remove-cell]
+import msprime
+import tskit
+
+def sim_example1():
+    ts = msprime.sim_ancestry(3, sequence_length=3e4,
+            recombination_rate=1e-8, population_size=1000,
+            record_full_arg=True, random_seed=21)
+    ts.dump("data/haplotypes1.trees")
+
+sim_example1()  # uncomment to recreate the tree seqs used in this notebook
+```
+
+```{code-cell} ipython3
+ts = tskit.load("data/haplotypes1.trees")
+ts.draw_svg(size=(400, 200), y_axis=True, time_scale='rank')
+```
+
+What we will do is to keep track of ancestrally inherited segments
+as we move back through time, splitting and coalescing these as we go.
+We will maintain the state as a dictionary mapping
+each node ID to a list of segments.
+What exactly does one of these lists of segments mean?
+Well, a node represents an ancestral genome that was present
+at a particular point in time.
+An edge represents a sequence of ancestral genomes along which
+a given segment was inherited.
+(Anoter interpretation would be that an edge represents
+a relationship: in other words, merely that a given node inherited
+a given chunk of genome from a given ancestral node,
+but let's go with the more concrete interpretation.)
+So, to initialize we need to start off with a segment
+that is the entire genome for each sample (since in this
+tree sequence, all samples are at time=0 ago).
+
+```{code-cell} ipython3
+ancestry = {n.id : SegmentList() for n in ts.nodes()}
+for j in ts.samples():
+    ancestry[j].add_segment(0, ts.sequence_length)
+```
+
+Now, edges in the EdgeTable are sorted by parent time,
+so if we iterate through the edges in order, we move back in time.
+So, we can use this to see the state of the process at, say,
+500 generations in the past:
+
+```{code-cell} ipython3
+for e in ts.edges():
+    t = ts.node(e.parent).time
```

Comment:
Maybe use ``ts.nodes_time[e.parent]`` so that that this is more easily translatable to numba?